### ISSUE TITLE
Google Calendar fails to load user.js due to newly implemented CSP "require-trusted-types-for 'script'"

### ISSUE THREAD

**Issue by Alexand3-R** [ferdium/ferdium-app#1086]
### Preflight Checklist

- [x] I agree to follow the [Code of Conduct](https://github.com/ferdium/ferdium-app/blob/HEAD/CODE_OF_CONDUCT.md) that Ferdium Recipes adheres to.
- [X] I have searched the [issue tracker](https://github.com/ferdium/ferdium-app/issues?q=is%3Aopen+is%3Aissue+label%3A%22recipe+%3Abookmark_tabs%3A%22+label%3A%22bug+%3Abeetle%3A%22+) for a bug report that matches the one I want to file, without success.

### Ferdium Version

6.2.4

### Recipe Name

Google Calendar

### Recipe Version

2.3.2

### Steps to reproduce

Simply check the Service Developers Tools Console for the following error:

`Uncaught (in promise) TypeError: Failed to set the 'innerHTML' property on 'Element': This document requires 'TrustedHTML' assignment.
    at v.loadUserFiles (VM133 recipe.js:1:4621)
    at v.loadRecipeModule (VM133 recipe.js:1:4526)
    at EventEmitter.<anonymous> (VM133 recipe.js:1:3337)
    at EventEmitter.emit (VM532 node:events:513:28)
    at Object.onMessage (VM639 renderer_init:2:9151)`

### Expected Behavior

Script should load fine and does load fine with other services that do not implement this specific CSP.

### Actual Behavior

The user.js script fails to load.

### Screenshots

The Service Developer Tools console shows the following error:
![afbeelding](https://user-images.githubusercontent.com/53404641/228579092-f6e341ff-d5cd-4063-9b86-af8a97ff935a.png)

### Additional Information

Also the default Hello World script fails to load.

**Comment by eitchtee** [comment]
Not sure if related, but `user.css` stopped loading today as well. Couldn't find a related error on console though.

**Comment by Julian88Tex** [comment]
Would this likely cause dark mode to not work? Just noticed that starting today.

**Comment by Alexand3-R** [comment]
> Would this likely cause dark mode to not work? Just noticed that starting today.

If its implemented using JavaScript, then yes, likely.
Possible fixes are:

**1. Modify Ferdium code and rebuild to have it load the user.js file using a policy. Something like this:**

```
const escapeHTMLPolicy = trustedTypes.createPolicy("myEscapePolicy", {
createScriptURL: (string) =>string.replace(/>/g, "<"),});
var newScript = document.createElement('script');
newScript.type = 'text/javascript';
newScript.src = escapeHTMLPolicy.createScriptURL(ferdiUserScriptPath);
```

**2. Find a way to fully disable CSP in Ferdium.**

Debugger output of error:
![afbeelding](https://user-images.githubusercontent.com/53404641/228706373-7506d2b4-64cd-4db3-9333-4642c450cf8d.png)

Afaik, regular calls like innerHTML are also stopped by this new CSP.

**Comment by Alexand3-R** [comment]
The code that likely needs to be fixed starts at line 260 in: ferdium/src/webview/recipe.ts

[https://github.com/ferdium/ferdium-app/blob/develop/src/webview/recipe.ts?plain=1#L260](https://github.com/ferdium/ferdium-app/blob/develop/src/webview/recipe.ts?plain=1#L260)

**Comment by Alexand3-R** [comment]
Here is a workaround that allows loading remote scripts. This should be appended to webview.js that resides in the recipe folder in order to work.

```
// Workaround for loading remote scripts
document.addEventListener('DOMContentLoaded', function() {
const remoteUserScript = "https://somedomain.com/path/to/script.js";
const loadScript = (FILE_URL, async = true, type = "text/javascript") => {
    return new Promise((resolve, reject) => {
        try {
        	var escapeHTMLPolicyA = trustedTypes.createPolicy("myEscapePolicy", {
	  			createScriptURL: (string) => string.replace(/>/g, "<"),
			});
            const scriptEle = document.createElement("script");
            scriptEle.type = type;
            scriptEle.async = async;
            scriptEle.src = escapeHTMLPolicyA.createScriptURL(FILE_URL);

            scriptEle.addEventListener("load", (ev) => {
                resolve({ status: true });
            });

            scriptEle.addEventListener("error", (ev) => {
                reject({
                    status: false,
                    message: `Failed to load the script ＄{FILE_URL}`
                });
            });

            document.body.appendChild(scriptEle);
        } catch (error) {
            reject(error);
        }
    });
};

loadScript(remoteUserScript)
    .then( data  => {
        console.log("Script loaded successfully", data);
    })
    .catch( err => {
        console.error(err);
    });
}, false);        
```

**Comment by Julian88Tex** [comment]
@Alexand3-R hm, gave that a try but my dark mode didn't come back :(

**Comment by lukassr** [comment]
I managed to bring back the dark mode. In some lines where the `innerHTML` is edited, I wrapped the new content with the following function:
```javascript
const escapeHTMLPolicy = trustedTypes.createPolicy('escapeHTML', {
  createHTML: string => string.replace(/</g, '&lt;'), // Escapes `<` characters to prevent the creation of new HTML elements
```
[Ref 1](https://developer.mozilla.org/en-US/docs/Web/API/TrustedHTML), [Ref 2](https://github.com/vercel/next.js/issues/13228#issuecomment-634250582)

Specifically, I used the function in the following lines:
- [https://github.com/ferdium/ferdium-app/blob/e9c6bb2c16f213e1b0c0a21a18711bef2ae96f3f/src/webview/recipe.ts#L262](https://github.com/ferdium/ferdium-app/blob/e9c6bb2c16f213e1b0c0a21a18711bef2ae96f3f/src/webview/recipe.ts#L262)
- [https://github.com/ferdium/ferdium-app/blob/e9c6bb2c16f213e1b0c0a21a18711bef2ae96f3f/src/webview/recipe.ts#L267](https://github.com/ferdium/ferdium-app/blob/e9c6bb2c16f213e1b0c0a21a18711bef2ae96f3f/src/webview/recipe.ts#L267)
- [https://github.com/ferdium/ferdium-app/blob/e9c6bb2c16f213e1b0c0a21a18711bef2ae96f3f/src/webview/lib/RecipeWebview.ts#L133](https://github.com/ferdium/ferdium-app/blob/e9c6bb2c16f213e1b0c0a21a18711bef2ae96f3f/src/webview/lib/RecipeWebview.ts#L133)

Dark mode in Google Calendar is back, but I'm having troubles disabling Dark Mode for some reason. Also, I didn't tested to see if those changes affect other recipes, but my eyes are burning and I use GCalendar in my day-to-day.

**Comment by Julian88Tex** [comment]
@lukassr thanks so much. Could you be specific about the steps to implement this in someone's local Ferdium app? My eyes are burning too!

**Comment by Alexand3-R** [comment]
> @Alexand3-R hm, gave that a try but my dark mode didn't come back :(

Sorry this fix is only for loading external JavaScript files. However, here is the workaround that allows loading remote CSS files. For this to work, you would have to upload your "darkmode.css" file somewhere that allows direct access. The darkmode file can be found in your recipes folder of googlecalendar. Again, append this to webview.js.

```
// Workaround for loading darkmode.css
var cssId = 'cssDarkModeWorkaround';
if (!document.getElementById(cssId))
{
    var head  = document.getElementsByTagName('head')[0];
    var link  = document.createElement('link');
    link.id   = cssId;
    link.rel  = 'stylesheet';
    link.type = 'text/css';
    link.href = 'https://somedomain.com/path/to/darkmode.css';
    link.media = 'all';
    head.appendChild(link);
}
```

**Comment by Alphrag** [comment]
@lukassr and @Alexand3-R would you be able to make a joint PR about this CSP issue please?

**Comment by Julian88Tex** [comment]
@Alexand3-R thanks for that. I'm still having trouble getting it to load in dark mode. This is what my webview.js file in my googlecalendar recipes folder looks like:

```
const _path = _interopRequireDefault(require('path'));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

module.exports = Ferdium => {
  Ferdium.injectCSS(_path.default.join(__dirname, 'calendar.css'));
  Ferdium.injectJSUnsafe(_path.default.join(__dirname, 'webview-unsafe.js'));
};

// Workaround for loading darkmode.css
var cssId = 'cssDarkModeWorkaround';
if (!document.getElementById(cssId))
{
    var head  = document.getElementsByTagName('head')[0];
    var link  = document.createElement('link');
    link.id   = cssId;
    link.rel  = 'stylesheet';
    link.type = 'text/css';
    link.href = 'https://raw.githubusercontent.com/ferdium/ferdium-recipes/main/recipes/googlecalendar/darkmode.css';
    link.media = 'all';
    head.appendChild(link);
}
```
Did I miss anything?

**Comment by Alexand3-R** [comment]
> 

Hi, for some reason, the CSS's stored on Github do not load. That was what I first tried. I was able to get it working by hosting the file on my own web server. Perhaps you can find another service to upload the darkmode.css.

**Comment by MosEisley34** [comment]
@Julian88Tex Alexand3-R's script worked using the following URL: https://cdn.statically.io/gh/ferdium/ferdium-recipes/main/recipes/googlecalendar/darkmode.css

**Comment by Julian88Tex** [comment]
@Alexand3-R @MosEisley34 ah thanks so much that worked great for me 💜

**Comment by lukassr** [comment]
@Alexand3-R great job! Can you make a PR in [ferdium-recipes](https://github.com/ferdium/ferdium-recipes) for this fix? I think this approach is less invasive than mine, because it only modifies the `webview.js` from the Google Calendar recipe. I could do it aswell as a joint PR, if you don't have time.

**Comment by Alexand3-R** [comment]
> 

Hi @lukassr, the issue with these workarounds are that they only allow loading remotely hosted scripts and stylesheets, since loading from disk is a security breach in browsers/electron. The real fix would be to alter Ferdium's code and ensure that the local files, the user.js and user.css are loaded using a trusted types policy to prevent CSP errors. The policy may even be empty. As long as a policy is used, it will work. I don't have an environment set up to make changes to the Ferdium application.

**Comment by lukassr** [comment]
@Alexand3-R I see. Sounds a little ugly but we can use the static URL from the GitHub ferdium-recipes repo to apply _more correctly_ the fix, as we maintain only files within the repo. Using [https://raw.githubusercontent.com/ferdium/ferdium-recipes/main/recipes/googlecalendar/darkmode.css](https://raw.githubusercontent.com/ferdium/ferdium-recipes/main/recipes/googlecalendar/darkmode.css) will work as a workaround for Google Calendar, as the link will always work. We can work in another issue for the final solution to prevent CSP errors across all Ferdium, but I see that the only recipe with troubles right now is Google Calendar. What do you think?

**Comment by Alexand3-R** [comment]
> @Alexand3-R I see. Sounds a little ugly but we can use the static URL from the GitHub ferdium-recipes repo to apply _more correctly_ the fix, as we maintain only files within the repo. Using https://raw.githubusercontent.com/ferdium/ferdium-recipes/main/recipes/googlecalendar/darkmode.css will work as a workaround for Google Calendar, as the link will always work. We can work in another issue for the final solution to prevent CSP errors across all Ferdium, but I see that the only recipe with troubles right now is Google Calendar. What do you think?

Using CSS files hosted on raw.githubusercontent.com does not work for some reason, however @MosEisley34 states that https://cdn.statically.io/gh/ferdium/ferdium-recipes/main/recipes/googlecalendar/darkmode.css does work, so that should be used instead.

However, also for the CSS, its not a workaround that is viable enough to be pushed to the googlecalendar recipe, as the options in the recipe settings to toggle it on and off and certain parameters, such as brightness, sepia and contrast will not work.

**Comment by th3jesta** [comment]
The workaround is good enough for me until a real fix gets pushed. Thanks, everyone! My corneas also thank you.

**Comment by mikeseese** [comment]
In https://github.com/ferdium/ferdium-recipes/pull/397, the directory got renamed from `googlecalendar` to `google-calendar`. Here is the updated `webview.js` for the mentioned workaround:

``` js
const _path = _interopRequireDefault(require('path'));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

module.exports = Ferdium => {
  // if the user is on googlecalendar landing page, go to the login page.
  if (
    location.hostname == 'workspace.google.com' &&
    location.href.includes('products/calendar/')
  ) {
    location.href =
      'https://accounts.google.com/AccountChooser?continue=https://calendar.google.com/u/0/';
  }
  Ferdium.injectCSS('https://cdn.statically.io/gh/ferdium/ferdium-recipes/main/recipes/google-calendar/calendar.css');
  Ferdium.injectJSUnsafe('https://cdn.statically.io/gh/ferdium/ferdium-recipes/main/recipes/google-calendar/webview-unsave.js');

  // Workaround for loading darkmode.css
  var cssId = 'cssDarkModeWorkaround';
  if (!document.getElementById(cssId))
  {
      var head  = document.getElementsByTagName('head')[0];
      var link  = document.createElement('link');
      link.id   = cssId;
      link.rel  = 'stylesheet';
      link.type = 'text/css';
      link.href = 'https://cdn.statically.io/gh/ferdium/ferdium-recipes/main/recipes/google-calendar/darkmode.css';
      link.media = 'all';
      head.appendChild(link);
  }
};

```

**Comment by Julian88Tex** [comment]
@mikeseese thanks so much for this updated code. This worked for me :)

**Comment by mcmxcdev** [comment]
This was fixed in https://github.com/ferdium/ferdium-recipes/pull/402

**Comment by tfluehmann** [comment]
This was closed as fixed by ferdium/ferdium-recipes#402, but I don't think the underlying bug ever went away — that PR routed the Google Calendar recipe's *own* stylesheet around the problem by loading it from a CDN, while the code that actually throws was left untouched.

`src/webview/recipe.ts` still does:

```js
const styles = document.createElement('style');
const userCss = join(recipe.path, 'user.css');
if (pathExistsSync(userCss)) {
  const data = readFileSync(userCss);
  styles.innerHTML += data.toString();   // <- Trusted Types sink
```

`innerHTML` is a TrustedHTML sink, so on any service sending `require-trusted-types-for 'script'` that assignment still throws — and because it escapes `loadUserFiles()` before reaching the `user.js` block further down the same function, one `user.css` still takes `user.js` down with it. That matches the `at v.loadUserFiles` frame in the original report exactly.

I verified in Electron 37 against a page with that CSP that the fix is simply `textContent` (not a sink, and equivalent for a `<style>` element):

```json
{
  "innerHTML":   "TypeError: Failed to set the 'innerHTML' property on 'Element': This document requires 'TrustedHTML' assignment.",
  "textContent": "OK",
  "appliedFontWeight": "700",           // textContent stylesheet took effect
  "appliedColor":      "rgb(0, 0, 0)"   // innerHTML stylesheet silently lost
}
```

There are three other call sites with the same sink (`darkmode.ts`, `RecipeWebview.ts`, `contextMenuBuilder.ts`), so `darkmode.css` and recipe `injectCSS()` are affected too. Opened #2463 with the fix.

No need to reopen this if you'd rather track it on the PR — just flagging that the "fixed" status is misleading for anyone searching for why their `user.css` is ignored.
